fili-core/src/main/java/com/yahoo/bard/webservice/web/responseprocessors/EtagCacheResponseProcessor.java

```diff
@@ -0,0 +1,137 @@
+import static javax.ws.rs.core.Response.Status.NOT_MODIFIED;
+import static javax.ws.rs.core.Response.Status.OK;
+
+import com.yahoo.bard.webservice.data.cache.TupleDataCache;
+import com.yahoo.bard.webservice.druid.client.FailureCallback;
+import com.yahoo.bard.webservice.druid.client.HttpErrorCallback;
+import com.yahoo.bard.webservice.druid.model.query.DruidAggregationQuery;
+import com.yahoo.bard.webservice.web.ErrorMessageFormat;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * A response processor which caches the results if appropriate after completing a query according to etag value.
+ */
+public class EtagCacheResponseProcessor implements FullResponseProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EtagCacheResponseProcessor.class);
+
+    private final ObjectMapper mapper;
+    private final ObjectWriter writer;
+    private final ResponseProcessor next;
+    private final String cacheKey;
+    private final TupleDataCache<String, String, String> dataCache;
+
+    /**
+     * Constructor.
+     *
+     * @param next  Next ResponseProcessor in the chain
+     * @param cacheKey  Key into which to write a cache entry
+     * @param dataCache  The cache into which to write a cache entry
+     * @param mapper  An object mapper to use for processing Json
+     */
+    public EtagCacheResponseProcessor(
+            @NotNull ResponseProcessor next,
+            @NotNull String cacheKey,
+            @NotNull TupleDataCache<String, String, String> dataCache,
+            @NotNull ObjectMapper mapper
+    ) {
+        this.next = next;
+        this.cacheKey = cacheKey;
+        this.dataCache = dataCache;
+        this.mapper = mapper;
+        this.writer = mapper.writer();
+    }
+
+    @Override
+    public ResponseContext getResponseContext() {
+        return next.getResponseContext();
+    }
+
+    @Override
+    public FailureCallback getFailureCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getFailureCallback(druidQuery);
+    }
+
+    @Override
+    public HttpErrorCallback getErrorCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getErrorCallback(druidQuery);
+    }
+
+    @Override
+    public void processResponse(JsonNode json, DruidAggregationQuery<?> druidQuery, LoggingContext metadata) {
+        // make sure JSON response comes with status code
+        if (!json.has(DruidJsonResponseContentKeys.STATUS_CODE.getName())) {
+            logAndGetErrorCallback(ErrorMessageFormat.STATUS_CODE_MISSING_FROM_RESPONSE.format(), druidQuery);
+            return;
```

Comment:
Well, if you assume `logAndGetErrorCallback` does `return`, it's actually not the case.